### PR TITLE
Stu3 complete export

### PR DESCRIPTION
**Pull request by kjmahalingam**
This PR implements full STU3 export functionality. Additional information is added to resources such as Encounter, MedicationRequest, Procedure, etc. 

This functionality still requires unit tests. This functionality will eventually need to be verified to ensure that it outputs valid STU3 FHIR.

### CODE REVIEW

**Comment by jawalonoski** [comment]
This introduced something like 1800 checkstyle warnings. Perhaps I should change the level to ERROR... because most of the changes are just style and indentation... I can't tell what has changed when you look at the diff.

File: src/main/java/org/mitre/synthea/export/FhirStu3.java [hunk 1]
```diff
@@ -86,19 +101,21 @@
...
   private static final String RXNORM_URI = "http://www.nlm.nih.gov/research/umls/rxnorm";
   private static final String CVX_URI = "http://hl7.org/fhir/sid/cvx";
+  private static final String DISCHARGE_URI = "http://www.nubc.org/patient-discharge";
   private static final String SHR_EXT = "http://standardhealthrecord.org/fhir/StructureDefinition/";
   private static final String SYNTHEA_EXT = "http://synthetichealth.github.io/synthea/";
 
   private static final Map raceEthnicityCodes = loadRaceEthnicityCodes();
   private static final Map languageLookup = loadLanguageLookup();
 
+
```
Comment (kjmahalingam) [inline src/main/java/org/mitre/synthea/export/FhirStu3.java]: Whitespace

File: src/main/java/org/mitre/synthea/export/FhirStu3.java [hunk 2]
```diff
@@ -346,60 +357,63 @@ private static BundleEntryComponent basicInfo(Person person, Bundle bundle, long
...
     patientResource.addExtension(birthSexExtension);
 
-    Extension mothersMaidenNameExtension = new Extension(
-        "http://hl7.org/fhir/StructureDefinition/patient-mothersMaidenName");
-    String mothersMaidenName = (String) person.attributes.get(Person.NAME_MOTHER);
+    Extension mothersMaidenNameExtension = new Extension("http://hl7.org/fhir/StructureDefinition/patient-mothersMaidenName");
+    String mothersMaidenName = (String)person.attributes.get(Person.NAME_MOTHER);
     mothersMaidenNameExtension.setValue(new StringType(mothersMaidenName));
     patientResource.addExtension(mothersMaidenNameExtension);
-    patientResource.addName().addGiven((String) person.attributes.get(Person.NAME));
```
Comment (kjmahalingam) [inline src/main/java/org/mitre/synthea/export/FhirStu3.java]: Removed redundant gender assignment

File: src/main/java/org/mitre/synthea/export/FhirStu3.java [hunk 3]
```diff
@@ -346,60 +357,63 @@ private static BundleEntryComponent basicInfo(Person person, Bundle bundle, long
...
-
-    if (person.attributes.get(Person.GENDER).equals("M")) {
```
Comment (kjmahalingam) [inline src/main/java/org/mitre/synthea/export/FhirStu3.java]: Removed redundant gender assignment

File: src/main/java/org/mitre/synthea/export/FhirStu3.java [hunk 4]
```diff
@@ -455,61 +472,115 @@ private static BundleEntryComponent encounter(BundleEntryComponent personEntry,
...
       encounterResource.addType(mapCodeToCodeableConcept(code, SNOMED_URI));
     }
 
     encounterResource.setClass_(new Coding().setCode(encounter.type));
-    long encounterEnd = encounter.stop > 0 ? encounter.stop
-        : encounter.start + TimeUnit.MINUTES.toMillis(15);
 
-    encounterResource
-        .setPeriod(new Period().setStart(new Date(encounter.start)).setEnd(new Date(encounterEnd)));
+    long encounterEnd =
```
Comment (kjmahalingam) [inline src/main/java/org/mitre/synthea/export/FhirStu3.java]: Implemented Encounter provider, reason, and discharge

File: src/main/java/org/mitre/synthea/export/FhirStu3.java [hunk 5]
```diff
@@ -734,8 +803,22 @@ private static BundleEntryComponent procedure(BundleEntryComponent personEntry,
       procedureResource.setPerformed(convertFhirDateTime(procedure.start, true));
     }
 
+    if (!procedure.reasons.isEmpty()) {
```
Comment (kjmahalingam) [inline src/main/java/org/mitre/synthea/export/FhirStu3.java]: Implement Procedure reason

File: src/main/java/org/mitre/synthea/export/FhirStu3.java [hunk 6]
```diff
@@ -786,33 +866,90 @@ private static BundleEntryComponent medication(BundleEntryComponent personEntry,
       medicationResource.setStatus(MedicationRequestStatus.ACTIVE);
     }
 
-    // TODO - prescription details & reason
+    if (!medication.reasons.isEmpty()) {
```
Comment (kjmahalingam) [inline src/main/java/org/mitre/synthea/export/FhirStu3.java]: Implement prescription details and reason

File: src/main/java/org/mitre/synthea/export/FhirStu3.java [hunk 7]
```diff
@@ -849,34 +981,202 @@ private static BundleEntryComponent careplan(BundleEntryComponent personEntry, B
...
     } else {
       careplanResource.setStatus(CarePlanStatus.ACTIVE);
-      // activityStatus = CarePlanActivityStatus.INPROGRESS;
-      // goalStatus = GoalStatus.INPROGRESS;
+      activityStatus = CarePlanActivityStatus.INPROGRESS;
+      goalStatus = GoalStatus.INPROGRESS;
     }
 
-    // TODO - goals, activities, reasons
+    if (!carePlan.activities.isEmpty()) {
```
Comment (kjmahalingam) [inline src/main/java/org/mitre/synthea/export/FhirStu3.java]: Implement CarePlan goals, activities, and reasons

File: src/main/java/org/mitre/synthea/export/FhirStu3.java [hunk 8]
```diff
@@ -849,34 +981,202 @@ private static BundleEntryComponent careplan(BundleEntryComponent personEntry, B
...
+
+    for (JsonObject goal : carePlan.goals) {
+      BundleEntryComponent goalEntry = caregoal(bundle, goalStatus, goal);
+      careplanResource.addGoal().setReference(goalEntry.getFullUrl());
+    }
 
     return newEntry(bundle, careplanResource);
   }
 
+  /**
```
Comment (kjmahalingam) [inline src/main/java/org/mitre/synthea/export/FhirStu3.java]: Implement provider

File: src/main/java/org/mitre/synthea/export/FhirStu3.java [hunk 9]
```diff
@@ -849,34 +981,202 @@ private static BundleEntryComponent careplan(BundleEntryComponent personEntry, B
...
+            "Healthcare Provider"));
+
+    organizationResource.setId(provider.getResourceID());
+    organizationResource.setName((String)provider.getAttributes().get("name"));
+    organizationResource.setType(organizationType);
+
+    return newEntry(bundle, organizationResource);
+  }
+
+  /**
```
Comment (kjmahalingam) [inline src/main/java/org/mitre/synthea/export/FhirStu3.java]: Implement caregoal

File: src/main/java/org/mitre/synthea/export/FhirStu3.java [hunk 10]
```diff
@@ -849,34 +981,202 @@ private static BundleEntryComponent careplan(BundleEntryComponent personEntry, B
...
+            }
+          }
+        }
+      }
+    }
+
+    return newEntry(bundle, goalResource);
+  }
+
+  /**
```
Comment (kjmahalingam) [inline src/main/java/org/mitre/synthea/export/FhirStu3.java]: Implement UnitsOfTime conversion function

File: src/main/java/org/mitre/synthea/export/FhirStu3.java [hunk 11]
```diff
@@ -849,34 +981,202 @@ private static BundleEntryComponent careplan(BundleEntryComponent personEntry, B
     Code code = carePlan.codes.get(0);
     careplanResource.addCategory(mapCodeToCodeableConcept(code, SNOMED_URI));
 
-    // CarePlanActivityStatus activityStatus;
```
Comment (kjmahalingam) [inline src/main/java/org/mitre/synthea/export/FhirStu3.java]: Uncomment CarePlan status functionality

File: src/main/java/org/mitre/synthea/export/FhirStu3.java [hunk 12]
```diff
@@ -849,34 +981,202 @@ private static BundleEntryComponent careplan(BundleEntryComponent personEntry, B
...
-    // GoalStatus goalStatus;
+    CarePlanActivityStatus activityStatus;
+    GoalStatus goalStatus;
 
     Period period = new Period().setStart(new Date(carePlan.start));
     careplanResource.setPeriod(period);
     if (carePlan.stop > 0L) {
       period.setEnd(new Date(carePlan.stop));
       careplanResource.setStatus(CarePlanStatus.COMPLETED);
-      // activityStatus = CarePlanActivityStatus.COMPLETED;
```
Comment (kjmahalingam) [inline src/main/java/org/mitre/synthea/export/FhirStu3.java]: Uncomment CarePlan status functionality

File: src/main/java/org/mitre/synthea/export/FhirStu3.java [hunk 13]
```diff
@@ -849,34 +981,202 @@ private static BundleEntryComponent careplan(BundleEntryComponent personEntry, B
...
-      // goalStatus = GoalStatus.ACHIEVED;
+      activityStatus = CarePlanActivityStatus.COMPLETED;
+      goalStatus = GoalStatus.ACHIEVED;
     } else {
       careplanResource.setStatus(CarePlanStatus.ACTIVE);
-      // activityStatus = CarePlanActivityStatus.INPROGRESS;
```
Comment (kjmahalingam) [inline src/main/java/org/mitre/synthea/export/FhirStu3.java]: Uncomment CarePlan status functionality

**Pull request closed without merging**